client/client.go

```diff
@@ -2,21 +2,164 @@ package client
+	"k8s.io/client-go/1.5/rest"
 )
 
 // Operator operators etcd clusters atop Kubernetes.
 type Operator interface {
-	// Create creates an etcd cluster with the given spec.
+	// Create creates an etcd cluster.
 	Create(ctx context.Context, name string, spec *spec.ClusterSpec) error
 
-	// Delete deletes the etcd cluster with given name
+	// Delete deletes the etcd cluster.
 	Delete(ctx context.Context, name string) error
 
 	// Update updates the etcd cluster with the given spec.
 	Update(ctx context.Context, name string, spec *spec.ClusterSpec) error
 
-	// Status gets the status of the etcd cluster with the given name.
-	Status(ctx context.Context, name string) (error, *spec.ClusterStatus)
+	// Get gets the etcd cluster information.
+	Get(ctx context.Context, name string) (*spec.EtcdCluster, error)
+
+	// List lists all the etcd clusters.
+	List(ctx context.Context) (*spec.EtcdClusterList, error)
+}
+
+type operator struct {
+	tprClient *rest.RESTClient
+	tprName   string
+	ns        string
+}
+
+func NewOperator(namespace string) (Operator, error) {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	configureClient(cfg)
+
+	tprclient, err := rest.RESTClientFor(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &operator{
+		tprClient: tprclient,
+		tprName:   controller.TPRName,
+		ns:        namespace,
+	}, nil
+
+}
+
+func (o *operator) Create(ctx context.Context, name string, cspec *spec.ClusterSpec) error {
+	cluster := &spec.EtcdCluster{
+		ObjectMeta: v1.ObjectMeta{
+			Name: name,
+		},
+		Spec: cspec,
+	}
+
+	err := o.tprClient.Post().
+		Resource(o.tprName).
+		Namespace(o.ns).
+		Body(cluster).
+		Do().Error()
+
+	return err
+}
+
+func (o *operator) Delete(ctx context.Context, name string) error {
+	return o.tprClient.Delete().
+		Resource(o.tprName).
+		Namespace(o.ns).Name(name).Do().Error()
+}
+
+func (o *operator) Update(ctx context.Context, name string, spec *spec.ClusterSpec) error {
+	for {
+		e, err := o.Get(ctx, name)
+		if err != nil {
+			return err
+		}
+
+		e.Spec = spec
+		var statusCode int
+
+		err = o.tprClient.Put().
+			Resource(o.tprName).
+			Namespace(o.ns).
+			Name(name).
+			Body(e).
+			Do().StatusCode(&statusCode).Error()
+
+		if statusCode == http.StatusConflict {
+			continue
+		}
+
+		return err
+	}
+}
+
+func (o *operator) Get(ctx context.Context, name string) (*spec.EtcdCluster, error) {
+	cluster := &spec.EtcdCluster{}
+
+	err := o.tprClient.Get().
+		Resource(o.tprName).
+		Namespace(o.ns).
+		Name(name).
+		Do().Into(cluster)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return cluster, nil
+}
+
+func (o *operator) List(ctx context.Context) (*spec.EtcdClusterList, error) {
+	clusters := &spec.EtcdClusterList{}
+
+	err := o.tprClient.Get().
+		Resource(o.tprName).
+		Namespace(o.ns).
+		Do().Into(clusters)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return clusters, nil
+}
+
+func configureClient(config *rest.Config) {
```

Comment:
i move the scheme registration to init. there should not be any race in this way. i do not want to create an instance since it is tedious to maintain it.